core/server/master/src/main/java/alluxio/master/file/activesync/ActiveSyncManager.java

```diff
@@ -192,48 +220,96 @@ private void launchPollingThread(long mountId, ExecutorService executorService,
-        launchPollingThread(mountId, mExecutorService, SyncInfo.INVALID_TXID);
-        initSync = true;
-      } else {
-        return false;
+  public boolean applyAndJournal(Supplier<JournalContext> context, AddSyncPointEntry entry)
+      throws ConnectionFailedException, IOException, InvalidPathException {
+    long mountId = entry.getMountId();
+    AlluxioURI uri = new AlluxioURI(entry.getSyncpointPath());
+    boolean launchPollingThread = false;
+    try {
+      launchPollingThread = startSyncInternal(uri);
+      context.get().append(Journal.JournalEntry.newBuilder().setAddSyncPoint(entry).build());
+    } catch (InvalidPathException e) {
+      LOG.info("Resolution of path {} failed", uri);
+      return false;
+    } catch (Throwable t) {
+      ProcessUtils.fatalError(LOG, t, "Failed to apply %s", entry);
+      throw t; // fatalError will usually system.exit
+    }
+
+    if (launchPollingThread) {
+      launchPollingThread(mountId, mExecutorService, SyncInfo.INVALID_TXID);
+      try {
+        if (Configuration.getBoolean(PropertyKey.MASTER_ACTIVE_UFS_SYNC_INITIAL_SYNC)) {
+          mFileSystemMaster.activeSyncMetadata(uri, null, getExecutor());
+        }
+      } catch (IOException e) {
+        LOG.warn("Network connection error causing initial sync to fail for {}", uri);
+        stopSync(uri);
+        throw new ConnectionFailedException("Add sync point"
+            + uri.toString() + "failed because of network error");
       }
     }
+    return true;
+  }
+
+  /**
+   * Apply removeSyncPoint entry and journal the entry.
+   *
+   * @param context journal context
+   * @param entry removeSyncPoint entry
+   * @return true if successfully applied and journaled
+   */
+  public boolean applyAndJournal(Supplier<JournalContext> context, RemoveSyncPointEntry entry)
+      throws InvalidPathException {
+    AlluxioURI uri = new AlluxioURI(entry.getSyncpointPath());
     try {
-      // Initial sync after adding a sync point
-      if (initSync && Configuration.getBoolean(PropertyKey.MASTER_ACTIVE_UFS_SYNC_INITIAL_SYNC)) {
-        mFileSystemMaster.activeSyncMetadata(syncPoint, null, getExecutor());
-      }
+      stopSync(uri);
+      context.get().append(Journal.JournalEntry.newBuilder().setRemoveSyncPoint(entry).build());
+      return true;
+    } catch (InvalidPathException e) {
+      LOG.info("Path resolution failed for path {}, and the exception is {}", uri, e);
+      throw e;
+      // Do not journal since none of the state is actually changed in the memory
     } catch (IOException e) {
-      LOG.warn("Network connection error causing initial sync to fail");
-      stopSync(syncPoint);
-      throw new ConnectionFailedException("Adding syncpoint"
-          + syncPoint.toString() + "failed because of network error");
+      LOG.info("IO exception encountered when removing {} as a sync point", uri);
+      // Do journal since the state has already changed in the memory
```

Comment:
The pattern we use for inode tree is

1. Figure out what state needs to change
2. Create an entry for the state change
3. Apply the state change (this should be very basic, e.g. updating a map or two)
4. Write the entry to the journal

By the time we reach step (3), we should have ruled out any possibility of throwing an exception. Anything that can throw an exception should be handled in step (1) before we do anything journal-related. It is a fatal error if an `applyAndJournal` method throws an exception. Also, things like launching asynchronous threads should happen outside of `applyAndJournal`. `applyAndJournal` should contain only state changes and then eventually journaling.

Is it possible to do error handling before reaching the state-change and journal code?

Reply:
One of the tricky point is that one of the state change is to call down to UFS to record the syncpoint, which always can have a IOException thrown.  I was able to refactor some of the error handling out of the state change, but this one is difficult to avoid.